src/Trains.NET.Rendering/LayerRenderer/TerrainRenderer.cs

```diff
@@ -0,0 +1,157 @@
+        public string Name => "Terrain";
+
+        public void Render(ICanvas canvas, int width, int height)
+        {
+            var contourLevels = GenerateListOfContourPointsForEachContourLevel();
+
+            foreach (var contourLevel in contourLevels.Keys)
+            {
+                var contourPoints = contourLevels[contourLevel];
+                DrawContourLine(canvas, contourPoints);
+            }
+        }
+
+        private void DrawContourLine(ICanvas canvas, List<ContourPoint> contourPoints)
+        {
+            foreach (var contourPoint in contourPoints)
+            {
+                var orderedList = OrderListByDistanceFromPoint(contourPoint, contourPoints);
+                var firstNearest = orderedList.Skip(1).First(); // The first element will be the original element since that will have distance zero
+                var secondNearest = orderedList.Skip(2).First();
+
+                DrawContourLineInViewport(canvas, contourPoint, firstNearest);
+                DrawContourLineInViewport(canvas, contourPoint, secondNearest);
+            }
+        }
+
+        private void DrawContourLineInViewport(ICanvas canvas, ContourPoint contourPoint1, ContourPoint contourPoint2)
+        {
+            var grid = new PaintBrush
+            {
+                Color = Colors.LightGray,
+                StrokeWidth = 1,
+                Style = PaintStyle.Stroke
+            };
+
+            var (col1, row1) = _pixelMapper.CoordsToViewPortPixels(contourPoint1.Column, contourPoint1.Row);
+            var (col2, row2) = _pixelMapper.CoordsToViewPortPixels(contourPoint2.Column, contourPoint2.Row);
+
+            canvas.DrawLine(col1, row1, col2, row2, grid);
+        }
+
+        private Dictionary<int, List<ContourPoint>> GenerateListOfContourPointsForEachContourLevel()
+        {
+            var topography = new Dictionary<int, List<ContourPoint>>();
+            foreach (var terrain in _terrainLayout)
+            {
+                var contourLevel = CalculateContourLevel(terrain.Altitude);
+                var contourPoints = topography.ContainsKey(contourLevel) ? topography[contourLevel] : new List<ContourPoint>();
+
+                var adjacentTerrainLookups = new List<Func<Terrain, Terrain>>
+                {
+                    _terrainLayout.GetAdjacentTerrainUp,
+                    _terrainLayout.GetAdjacentTerrainDown,
+                    _terrainLayout.GetAdjacentTerrainLeft,
+                    _terrainLayout.GetAdjacentTerrainRight,
+                };
+
+                foreach (var adjacentTerrainLookup in adjacentTerrainLookups)
+                {
+                    var adjacentContourPoint = CalculateBorderingContourPointIfAdjacentTerrainIsLower(adjacentTerrainLookup, terrain, contourLevel);
+                    if (adjacentContourPoint != null && !contourPoints.Contains(adjacentContourPoint.Value))
+                        contourPoints.Add(adjacentContourPoint.Value);
```

Comment:
Braces around if statements or I'll not threaten you in public :D

Reply:
Done